vip-jetpack/connection-pilot/class-jetpack-connection-attendant.php

```diff
@@ -0,0 +1,309 @@
+		$legacy_owner_details   = $this->get_legacy_owner_details();
+		$org_owner_details      = $this->get_org_owner_details();
+
+		$current_owner = $this->get_current_connection_owner();
+		if ( ! isset( $current_owner->user_login ) ) {
+			return 'not_vip';
+		}
+
+		// We won't consider it the legacy user if it is still the expected owner.
+		if ( $expected_owner_details['login'] === $current_owner->user_login ) {
+			return 'is_vip';
+		}
+
+		if ( $legacy_owner_details && $legacy_owner_details['login'] === $current_owner->user_login ) {
+			return 'is_legacy_vip';
+		}
+
+		return 'not_vip';
+	}
+
+	/**
+	 * Returns the user object of the connection owner.
+	 *
+	 * @return WP_User|null Null if no connection owner found.
+	 */
+	private function get_current_connection_owner() {
+		if ( ! method_exists( 'Jetpack', 'connection' ) ) {
+			return null;
+		}
+
+		$jp_connection = \Jetpack::connection();
+
+		if ( ! method_exists( $jp_connection, 'get_connection_owner' ) ) {
+			return null;
+		}
+
+		$connection_owner = $jp_connection->get_connection_owner();
+		if ( false === $connection_owner ) {
+			return null;
+		}
+
+		return $connection_owner;
+	}
+
+	/**
+	 * Ensures the VIP user exists on the site and has the necessary permissions.
+	 *
+	 * @return WP_User|WP_Error WP_User if successful, WP_Error otherwise.
+	 */
+	public function ensure_user_exists() {
+		$details = $this->get_expected_owner_details();
+
+		$user = get_user_by( 'login', $details['login'] );
+
+		// Create user if it doesn't exist yet.
+		if ( ! $user ) {
+			$result = wp_insert_user( [
+				'user_login'   => $details['login'],
+				'user_email'   => $details['email'],
+				'display_name' => $details['display_name'],
+				'role'         => $details['role'],
+				'user_pass'    => wp_generate_password( 100 ),
+			] );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-creation-failed', 'Failed to create new user.' );
+			}
+
+			$user = get_userdata( $result );
+			if ( ! $user ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-creation-failed', 'Failed to create new user.' );
+			}
+		}
+
+		// Ensure the email is correct & hasn't been changed.
+		if ( $details['email'] !== $user->user_email ) {
+			$result = wp_update_user( [
+				'ID'         => $user->ID,
+				'user_email' => $details['email'],
+			] );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-update-failed', 'Failed to correct the email address on the primary user.' );
+			}
+		}
+
+		// Add user to blog if needed, and ensure they are a super admin.
+		if ( is_multisite() ) {
+			$blog_id = get_current_blog_id();
+
+			if ( ! is_user_member_of_blog( $user->ID, $blog_id ) ) {
+				$added_to_blog = add_user_to_blog( $blog_id, $user->ID, $details['role'] );
+
+				if ( is_wp_error( $added_to_blog ) ) {
+					return new WP_Error( 'jp-cxn-pilot-owner-ms-add-failed', 'Failed to add user to blog.' );
```

Comment:
Should we pass through the explanation from the WP_Error? Looking at code it can be either `user_does_not_exist` or user_cannot_be_added (based on a filter return value)

Reply:
Passing through the error code now 👍